src/ice_nucleation.F90

```diff
@@ -0,0 +1,513 @@
+    deallocate(H2O_masses)
+    deallocate(H2O_frac)
+
+  end subroutine ice_nucleation_immersion_freezing_time_dependent_naive
+
+!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+
+  !> This subroutine simulates the melting process.(Set frozen = .False.
+  !> for each particle once the temperature is higher than water freezing
+  !> tempearture)
+  subroutine ice_nucleation_melting(aero_state, aero_data, env_state)
+
+    !> Aerosol state.
+    type(aero_state_t), intent(inout) :: aero_state
+    !> Aerosol data.
+    type(aero_data_t), intent(in) :: aero_data
+    !> Environment state.
+    type(env_state_t), intent(inout) :: env_state
+
+    integer :: i_part
+
+    if (env_state%temp > const%water_freeze_temp) then
+       do i_part = 1, aero_state_n_part(aero_state)
+         aero_state%apa%particle(i_part)%frozen = .false.
+         aero_state%apa%particle(i_part)%den_ice = const%nan
+         aero_state%apa%particle(i_part)%ice_shape_phi = const%nan
+       end do
+    end if
+
+  end subroutine ice_nucleation_melting
+
+!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+
+  !> Calculating the freezing probability for the particle (i_part) using ABIFM
+  !> method (Knopf et al.,2013)
+  real(kind=dp) function ABIFM_Pfrz_particle(aero_particle, aero_data, &
+       a_w_ice, del_t)
+
+    !> Aerosol particle.
+    type(aero_particle_t), intent(in) :: aero_particle
+    !> Aerosol data.
+    type(aero_data_t), intent(in) :: aero_data
+    !> The water activity w.r.t. ice.
+    real(kind=dp), intent(in) :: a_w_ice
+    !> Time interval.
+    real(kind=dp), intent(in) :: del_t
+
+    real(kind=dp) :: aerosol_diameter
+    real(kind=dp) :: immersed_surface_area
+    real(kind=dp) :: total_vol
+    real(kind=dp) :: surface_ratio
+    real(kind=dp) :: abifm_m, abifm_c
+    real(kind=dp) :: j_het, j_het_x_area
+    integer :: i_spec
+
+    aerosol_diameter =  aero_particle_dry_diameter(aero_particle, aero_data)
+    immersed_surface_area = const%pi * aerosol_diameter **2
+
+    total_vol = 0d0
+    total_vol = aero_particle_dry_volume(aero_particle, aero_data)
+
+    j_het_x_area = 0d0
+    do i_spec = 1,aero_data_n_spec(aero_data)
+       if (i_spec == aero_data%i_water) cycle
+       abifm_m = aero_data%abifm_m(i_spec)
+       abifm_c = aero_data%abifm_c(i_spec)
+       j_het = 10 ** (abifm_m  * (1 - a_w_ice) + abifm_c) * 10000
+       surface_ratio = aero_particle%vol(i_spec) / total_vol
+       j_het_x_area = j_het_x_area + j_het * immersed_surface_area * &
+            surface_ratio
+    end do
+
+    ABIFM_Pfrz_particle = 1 - exp(-j_het_x_area * del_t)
+
+  end function ABIFM_Pfrz_particle
+
+!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+
+  !> Calculating the heterogeneous ice nucleation rate coefficient
+  !> for each species, and fining the species having the largest rate.
+  subroutine ABIFM_max_spec(aero_data, a_w_ice, i_spec_max, j_het_max)
```

Comment:
Is there a reason that we have to know what species it is? it appears we only do something with the maximum rate.

I'd shorten the description a bit to what its purpose is which is finding the maximum heterogeneous ice nucleation rate coefficient.